docs/how_to_use_adu_client.md

```diff
@@ -0,0 +1,166 @@
+
+    If the update request is for a new update and the device (ADU Agent) decides to accept it, call `AzureIoTADUClient_SendAgentState` with a `eAzureIoTADUAgentStateDeploymentInProgress` state to inform the ADU Service that the update is in progress.
+
+1. Download update image(s) and write to Flash memory.
+
+    The device application must iterate through the files in the update request, download them and write them to flash memory as applicable.
+
+    The links to the binaries to be downloaded are stored in `AzureIoTADUUpdateRequest_t.pxFileUrls[]`.
+    Each downloaded file shall be validated against its hash present in `AzureIoTADUUpdateRequest_t.xUpdateManifest.pxFiles[].pxHashes[]`.
+
+    The Azure IoT Middleware for FreeRTOS provides an interface implemented on top of FreeRTOS' coreHTTP library to help download the binary files referenced in the update request. Header [azure_iot_http.h](https://github.com/Azure/azure-iot-middleware-freertos/blob/main/source/interface/azure_iot_http.h) exposes the following functions:
+    - `AzureIoTHTTP_RequestSizeInit`
+    - `AzureIoTHTTP_RequestSize`
+    - `AzureIoTHTTP_Init`
+    - `AzureIoTHTTP_Request`
+    - `AzureIoTHTTP_Request`
+
+    See `prvDownloadUpdateImageIntoFlash` in [sample_azure_iot_adu.c](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/main/demos/sample_azure_iot_adu/sample_azure_iot_adu.c) for an example how to use these functions to download and write images to flash memory.
+
+    > If an update fails (e.g., downloading the image files, or writing to flash), `AzureIoTADUClient_SendAgentState` must be called twice; once with state `eAzureIoTADUAgentStateFailed`, followed by another call with state `eAzureIoTADUAgentStateIdle` (with the same image version as before the update request).
+
+1. Reboot device and load new image.
+
+    At this point the device application must set the device to load the new image version upon reboot.
+    This usually involves setting the bank or partition with the new image version (just downloaded) as active/enabled.
+
+    Once that is done, the device application must restart and load the new image version.
+
+    See `prvEnableImageAndResetDevice` in [sample_azure_iot_adu.c](https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/main/demos/sample_azure_iot_adu/sample_azure_iot_adu.c) .
+
+
+1. Send the ADU Agent state to the ADU Service.
+
+    Once the device update is successfuly completed the device application must inform the ADU Agent state to the ADU Service.
+
+    This is also done with `AzureIoTADUClient_SendAgentState`.
+    It will inform the ADU Service which device this is (manufacturer, model), the version of its new image (if successful) and that this device is ready to receive updates (state `eAzureIoTADUAgentStateIdle`).
+    
+
+### Flowchart of ADU Agent Logic
+
+<img src="./resources/adu-agent-logic.jpg" width="100%">
```

Comment:
note this is missing a step to verify the signature I believe.

Reply:
Ohh. You are right. Mermaid will help here. Thanks @CIPop

Reply:
Fixed.